HMNFD_Design.md

```diff
@@ -0,0 +1,274 @@
+"allnodes".  This greatly simplifies and speeds up subscription matching
+if the subscriber can tolerate it.
+
+In addition to nodes doing SCN subscriptions, HMNFD itself will subscribe
+for all possible SCNs from HSM.
+
+#### Pruning
+
+The API provides means to generate SCN subscriptions as well as delete
+them.  In addition, if HMNFD sees that a given node or component has
+"gone away" (e.g. receiving an SCN for a component stating it is now "OFF"),
+then HMNFD will prune all of that node's subscriptions.
+
+Also, if an SCN comes in and is to be delivered to a node, and that node
+gets errors indicating it no longer exists, its subscriptions will
+also be pruned.
+
+### SCN Reception And Fanout
+
+HSM generates all SCNs.  All SCNs are sent to HMNFD via round-robin 
+distribution to the multiple running HMNFD instances.  
+
+Once an SCN is received by HMNFD, it will consult ETCD to get subscription
+records.  Once these are retrieved (in bulk for speed), they are matched
+against the SCN parameters to determine which nodes need to receive the
+SCN.
+
+The contents of the SCN are also examined to look for any subscribers that
+are no longer alive, and if any are found, those subscriptions are removed.
+
+The SCN is then sent to all matching subscribers using a goroutine worker
+pool to help parallelize things.
+
+When sending SCNs to subscribers, multiple attempts are made on failure,
+using a back-off algorithm for best results.  As stated earlier, if
+an SCN cannot be delivered after all retries are exhausted, SCN will
+delete that subscriber's subscription record and no further SCN delivery
+attempts will be made.
+
+#### SCN Batching
+
+To minimize SCN traffic, HMNFD will batch up the target subscribers
+in SCNs that match all other subscription parameters.  Thus, rather
+than sending an SCN to all subscribers for each one that is received
+from HSM, fewer SCNs with larger component lists will be sent.
+
+SCNs are batched for a maximum length of time, or maximum component
+count, whichever is less.  Once a batched SCN is ready to send, it
+is sent to all subscribers.
+
+This greatly cuts down on the number of messages sent, since most SCNs
+happen during batch node power-ups and boot operations, which tend
+to send lots of identical SCNs for many components in large bursts.
+Batching reduces the SCNs sent to one (or very few) per burst rather
+than one per individual SCN.
+
+
+### SCN Distribution To The SMA Framework
+
+When SCNs are received  by HMNFD, they are placed on the SMA Kafka bus
+and end up in the SMA database.  This allows for visualization tools 
+like Grafana/Kibana/etc. to view component state changes over time.
+
+### Health, Liveness, and Readiness
+
+HMNFD has health, liveness, and readiness APIs which are used by Kubernetes
+to determine the fitness of HMNFD.  Various things are checked, such as
+ETCD connectivity, etc. to determine fitness and health.
+
+## hmnfd API
+
+_hmnfd_'s RESTful API is a follows:
```

Comment:
if all of this detail is in the swagger I think we should point at the swagger and remove it from here, that way there is only 1 place to keep up.

Reply:
Done.